src/routing_table/mod.rs

```diff
@@ -714,13 +714,19 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
             return BTreeSet::new();
         }
 
-        self.merge(&merge_details.prefix);
-
-        // Establish list of provided contacts which are currently missing from our table.
-        merge_details.section
-            .difference(unwrap!(self.sections.get(&merge_details.prefix)))
-            .cloned()
-            .collect()
+        if self.sections.keys().any(|prefix| {
+            prefix.is_compatible(&merge_details.prefix) &&
+            prefix.bit_count() >= merge_details.prefix.bit_count()
+        }) {
+            self.merge(&merge_details.prefix);
+            // Establish list of provided contacts which are currently missing from our table.
+            merge_details.section
+                .difference(unwrap!(self.sections.get(&merge_details.prefix)))
+                .cloned()
+                .collect()
+        } else {
+            BTreeSet::new()
```

Comment:
I guess a comment like "We've already handled this merge." would be helpful here.